---
title: "Prepare Data"
output: html_document
date: '2023-07-28'
---

```{r setup, include=FALSE}
## using a development version of Spaniel
#devtools::install_github("TaSB-ncl/Spaniel",
#                         ref = "bioconductor3.18")                   
#

library(Spaniel)

## Bioconductor packages
library(SpatialExperiment)
library(SpatialFeatureExperiment)
library(SingleCellExperiment)
library(scater)
library(scuttle)
library(scran)
library(bluster)

## Cran Packages
library(tidyverse)
library(tidyterra)
library(rjson)
library(patchwork)

theme_set(theme_bw())

```


## Sample Metadata

The data being used in this workflow is taken from "Spatial Gene Expression 
Changes in the Mouse Heart After Base-Targeted Irradiation"

Radiation cardiotoxicity (RC) is a clinically significant 
adverse effect of treatment for patients with thoracic malignancies. 
Clinical studies in lung cancer have indicated that heart substructures 
are not uniformly radiosensitive, and that dose to the heart base drives RC.

This was a pilot experiment to determine whether spatial transcriptomics 
could provide useful insights in this context. 


### Experimental Design:
2 samples
2 conditions - irradiated heart and control
4 adjacent sections taken from each tissue block (sample)
8 sections in total


The sample metadata is summarised below:


```{r}
dataDir <- "testData/results"

samples <- list.files(dataDir)


samples_short <- samples %>% 
  gsub("results_", "", .)  %>% 
  gsub("_Results", "", .)

sample_id <- samples_short %>% 
  gsub("_.*", "", .)  



condition <- c(rep("Control", 2), rep("irradiated", 2))
tissue <- c(c(rep("tissue1", 2), rep("tissue2", 2)))

sampleInfo <- data.frame(Fname = samples, 
          id = samples_short, 
          condition, 
          tissue, 
          sample_id
          )

sampleInfo


pathToTenXOuts <- file.path(dataDir, sampleInfo$Fname, "outs")
```

## Import the expression data

The SpatialFeatureExperiment library contains an import function for 10X data which takes the path to 10X files and loads the data in as a SpatialFeatureExperiment object

```{r, eval=FALSE}
# sfe <- SpatialFeatureExperiment::read10xVisiumSFE(
#                                       dirs = pathToTenXOuts, 
#                                       samples = sampleInfo$Fname, 
#                                       sample_id = sampleInfo$sample_id, 
#                                       type = "sparse", 
#                                       data = "filtered", 
#                                       images = "hires",
#                                       zero.policy=TRUE)
# saveRDS(sfe, "rObjects/sfe.rds")
```

The import function is a bit time consuming so save time I have pre-computed a 
loaded rds file to work from:

```{r}
sfe <- readRDS("rObjects/sfe.rds")
```


## Add metadata and spot positions to data

This function adds some QC metadata to the object, 
as well as the spot coordinates which are used for plotting. For more details 
the functions are included 
R/sfe_prepareSFE.R


```{r}
sfe <- Spaniel::prepareSFE(sfe, sampleInfo)
```


## Add Image annotations

Different regions of the heart were annotated using the imaging software Amira. These could also be created using Photoshop or other imaging software. Each annotation was then exported as a separate layer. The annotation images are the same size as the high resolution images exported from 

For example this image shows the left ventricle in sample "C01":
![Left Ventricle sample C01](../testData/annotation_images/C01_left-ventricle.jpg)

The annotation_images directory 


```{r}
annotationDir <- "testData/annotation_images/"
annotations <- list.files(annotationDir) 
annoInfo <- data.frame(sample_id = gsub("_.*", "", annotations) ,
  domain = gsub(".*_", "", annotations) %>% 
    gsub("\\.jpg", "", .),jpg = annotations) 
```

## Add radiation field

![Radiation field = black rectangle; right atrium = yellow; left atrium = blue; great vessels = red; right ventricle = purple; left ventricle = green; atrioventricular valves = orange; central fibrous body = black; irradiated right ventricle = white; irradiated left ventricle = pink.](https://www.redjournal.org/cms/attachment/4177ae15-38f6-4470-b868-f5a460edf7ec/gr5.jpg)


```{r}
annoInfo$radiation_field <- ifelse(annoInfo$domain %in% c("left-ventricle", 
                                                          "right-ventricle"), 
                                   yes = FALSE, no = TRUE)
```

```{r}
annoInfo
```

The annotations can be loaded in and added to the sfe object using the spaniel function findAllDomains. This creates a separate for each histological domain in the colData of the object. The combineDomains function can then be used to combine the annotations into a single column named domain.

These functions are found in: R/sfe_findDomain.R

```{r}
sfe <- Spaniel::findAllDomains(sfe, annoInfo, annotationDir)
## create a column containing all domains
sfe <- combineDomains(sfe, unique(annoInfo$domain))

sfe$domain

## add radiation field
sfe$radiation_field <- colData(sfe) %>% 
  data.frame() %>% 
  left_join(annoInfo) %>% 
  pull(radiation_field)

```




## The domains can now be visualised on the histological data

```{r}
p1 <-  Spaniel::spanielPlot_SFE(sfe,  
                 sample_id = "C01", 
                 plot_feat = "domain", ptSizeMin = 0, 
                 ptSizeMax = 2) 

p2 <-  Spaniel::spanielPlot_SFE(sfe,  
                 sample_id = "C01", 
                 plot_feat = "radiation_field", ptSizeMin = 0, 
                 ptSizeMax = 2) 
p1 + p2

```

## Sample QC


Now that the image annotations are fully loaded in the transcriptomics data can be analysed.

QC metrics about each spot can be calculated using the addPerCellQCMetrics function from scuttle package. By default the total genes per spot are calculated. Additional metrics such as the percentage of mitochondrial genes can be calculated. As this data is from mouse mitochondrial genes start with the prefix "mt-" and can be extracted based on pattern matching.

```{r}
is_mt <- str_detect(rowData(sfe)$symbol, "^mt-")
sfe <- scuttle::addPerCellQCMetrics(sfe, subsets = list(mito = is_mt))
```

Spots where no genes are detected can be removed from the remainder of the analysis.

```{r}
filter <- sfe$detected > 0 & sfe$in_tissue
sfe <- sfe[, filter]
```



## Split the data into control and irradiated

Although we could now progress with the analysis using all samples it is often advisable to analyse each tissue sample or condition separately to begin with. This allows you to identify any section or sample which might be an outlier and should be removed from downstream analysis. For this workshop we example we will split the samples by condition:

```{r}
sfe_control <- sfe[,sfe$condition == "Control"]
sfe_irradiated <- sfe[,sfe$condition == "irradiated"]

```


## Cluster analysis


The filtered data can then be then be normalised and dimension reduction steps applied and the spots can be clustered according to transcriptomic similarity.  This analysis uses methods taken from the bioconductor Orchestrating Single-Cell Analysis https://bioconductor.org/books/release/OSCA/ work flows. 
Wrapper function (R/sfe_helper_functions.R) to perform this analysis are implemented in Spaniel for convenience. However for a more comprehensive clustering analysis we recommend following the reading the Scran and Bluster documentation which describes each individual step in the analysis in more detail.

In this example we use a Leidan clustering algorithm as implemented in scran. Leidan is a community detection method. It is important to note that the K parameter is the number of nearest neighbors to use when creating the k nearest neighbor graph. k is related to the resolution of the clustering result, a bigger k will result in lower resolution and vice versa.

```{r}
sfe_control <- sfe_control %>% 
  Spaniel::preProcess() %>%
  Spaniel::clustCells(K = 100, cluster_function = "leiden")
sfe_irradiated <- sfe_irradiated %>% 
   Spaniel::preProcess() %>% 
   Spaniel::clustCells(K = 100, cluster_function = "leiden")
```

## Control data

Now we can look at the results of the data both by visualising the clusters on a UMAP plot using plot functions from the scater package and within the spatial context of the tissue section using the spanielPlot function (R/sfe_PlotFunctions.R). 

```{r}
scater::plotUMAP(sfe_control, colour_by = "domain")
p1 <- Spaniel::spanielPlot_SFE(sfe_control,  
                 sample_id = "C01", 
                 plot_feat = "domain", ptSizeMin = 0, 
                 ptSizeMax = 2) 

p2 <- Spaniel::spanielPlot_SFE(sfe_control,  
                 sample_id = "C01", 
                 plot_feat = "clust_leiden_K_100", 
                 ptSizeMin = 0, 
                 ptSizeMax = 2) 

p1 + p2

```


## irradiated data

```{r}
scater::plotUMAP(sfe_irradiated, colour_by = "domain")
p1 <- Spaniel::spanielPlot_SFE(sfe_irradiated,  
                 sample_id = "C06", 
                 plot_feat = "domain", ptSizeMin = 0, 
                 ptSizeMax = 2) 

p2 <- Spaniel::spanielPlot_SFE(sfe_irradiated,  
                 sample_id = "C06", 
                 plot_feat = "clust_leiden_K_100", 
                 ptSizeMin = 0, 
                 ptSizeMax = 2) 
p1 + p2
```


## Save the processed data

```{r, eval = FALSe}
# saveRDS(sfe_control, "rObjects/wrapped_sfe_control.rds")
# saveRDS(sfe_irradiated, "rObjects/wrapped_sfe_irradiated.rds")
```


## Prepare the data to share

Finally we prepare the data so that it can be loaded into the Shiny Spaniel app. 
To demonstrate this in the interest of time are going to look at a subset of the data. 

```{r}
sfe_single <- sfe_control[,sfe_control$sample_id == "C01"]
```



The app allows us to look at QC features, and browse between different clustering solutions as well as different sample conditions.
Before we share our analysis with the rest of the team we can mark these columns in the metadata using the spaniel helper function "labelSFE".

We can check what we have in the metadata using the following function:

```{r}
colnames(colData(sfe_single))
```

Then we can define the columns that we want to label in our data for QC plots, groups (clusters or anatomical domains, histological regions) and the experimental conditions in our experiment. 

```{r}
grps <- c("clust_leiden_K_100", "domain", "radiation_field")
qc <- c("total", "sum", "subsets_mito_percent", "in_tissue")
exp_cond <- c("condition")

sfe_single <- Spaniel::labelSFE(sfe_single, qc_cols = qc, grp_cols = grps, exp_cond_cols = exp_cond)
```

As you can see the colData columns are now updated:


```{r}
colnames(colData(sfe_single))
```


```{r, eval = FALSE}
#saveRDS(sfe_single, "rObjects/sfe_single.rds")
```


## Run Shiny Spaniel!

Now we can launch the shiny app directly from rstudio

![Run shiny spaniel](shiny_spaniel.jpg)


```{r}
Spaniel::runShinySpaniel_SFE()
```



